src/modm/driver/encoder/as5047.hpp

```diff
@@ -0,0 +1,152 @@
+// coding: utf-8
+// ----------------------------------------------------------------------------
+/*
+ * Copyright (c) 2024, Henrik Hose
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#ifndef MODM_AS5047_HPP
+#define MODM_AS5047_HPP
+
+#include <array>
+#include <modm/architecture/interface/register.hpp>
+#include <modm/architecture/interface/spi_device.hpp>
+#include <modm/processing/resumable.hpp>
+#include <modm/processing/timer.hpp>
+#include <numbers>
+
+namespace modm
+{
+
+constexpr uint16_t
+setMsbToEvenParity(const uint16_t num)
+{
+	uint16_t par = 0x7fff & num;
+	par ^= par >> 8;
+	par ^= par >> 4;
+	par ^= par >> 2;
+	par ^= par >> 1;
+	return ((par & 1) << 15) | (0x7fff & num);
+}
+
+/// @ingroup modm_driver_max31865
+struct as5047
+{
+	enum class Errorfl : uint16_t
+	{
+		Parerr = Bit2,
+		Invcomm = Bit1,
+		Frerr = Bit0,
+	};
+	MODM_FLAGS16(Errorfl)
+
+	enum class Prog : uint16_t
+	{
+		Progver = Bit6,
+		Progotp = Bit3,
+		Otpref = Bit2,
+		Progen = Bit0,
+	};
+	MODM_FLAGS16(Prog)
+
+	enum class Diaagc : uint16_t
+	{
+		Magl = Bit11,
+		Magh = Bit10,
+		Cof = Bit9,
+		Lf = Bit8,
+	};
+	MODM_FLAGS16(Diaagc)
+
+	enum class Register : uint16_t
+	{
+		ReadNop = setMsbToEvenParity(((1 << 14) | 0x0000)),
+		ReadErrfl = setMsbToEvenParity(((1 << 14) | 0x0001)),
+		ReadProg = setMsbToEvenParity(((1 << 14) | 0x0003)),
+		ReadDiaagc = setMsbToEvenParity(((1 << 14) | 0x3FFC)),
+		ReadMag = setMsbToEvenParity(((1 << 14) | 0x3FFD)),
+		ReadAngleunc = setMsbToEvenParity(((1 << 14) | 0x3FFE)),
+		ReadAnglecom = setMsbToEvenParity(((1 << 14) | 0x3FFF)),
```

Comment:
You could move the `(1 << 14)` into the function.

Reply:
I don't think so, see [Figure 13 from the Datasheet](https://ams.com/documents/20143/36005/AS5047P_DS000324_3-00.pdf), bit 14 is the "READ/WRITE" bit, since these registers are only ever read, its unrelated to the parity bit.
I thought its more expressive this way, since it closely models what the datasheet says.
<details>
![image](https://github.com/modm-io/modm/assets/60779091/2e6076a1-8a8e-4592-8aa1-9251dddd3180)
</details>

Reply:
Yup that makes sense.